examples/reshare/src/application/core.rs

```diff
@@ -0,0 +1,106 @@
+//! Core reshare [Application] implementation.
+
+use crate::{
+    application::{genesis_block, Block},
+    dkg,
+};
+use commonware_consensus::{
+    marshal::ingress::mailbox::AncestorStream,
+    simplex::{signing_scheme::Scheme, types::Context},
+    Block as _, VerifyingApplication,
+};
+use commonware_cryptography::{
+    bls12381::primitives::variant::Variant, Committable, Digestible, Hasher, Signer,
+};
+use commonware_runtime::{Clock, Metrics, Spawner};
+use futures::StreamExt;
+use rand::Rng;
+use std::marker::PhantomData;
+
+#[derive(Clone)]
+pub struct Application<E, S, H, C, V>
+where
+    E: Rng + Spawner + Metrics + Clock,
+    S: Scheme,
+    H: Hasher,
+    C: Signer,
+    V: Variant,
+{
+    dkg: dkg::Mailbox<H, C, V>,
+    _marker: PhantomData<(E, S)>,
+}
+
+impl<E, S, H, C, V> Application<E, S, H, C, V>
+where
+    E: Rng + Spawner + Metrics + Clock,
+    S: Scheme,
+    H: Hasher,
+    C: Signer,
+    V: Variant,
+{
+    pub fn new(dkg: dkg::Mailbox<H, C, V>) -> Self {
+        Self {
+            dkg,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<E, S, H, C, V> commonware_consensus::Application<E> for Application<E, S, H, C, V>
+where
+    E: Rng + Spawner + Metrics + Clock,
+    S: Scheme,
+    H: Hasher,
+    C: Signer,
+    V: Variant,
+{
+    type Block = Block<H, C, V>;
+    type SigningScheme = S;
+    type Context = Context<H::Digest, C::PublicKey>;
+
+    async fn genesis(&mut self) -> Self::Block {
+        genesis_block::<H, C, V>()
+    }
+
+    async fn build(
+        &mut self,
+        _r_context: E,
+        _context: Self::Context,
+        mut ancestry: AncestorStream<Self::SigningScheme, Self::Block>,
+    ) -> Option<Self::Block> {
+        // Fetch the parent block from the ancestry stream.
+        let parent_block = ancestry.next().await?;
```

Comment:
While this is clean, it may not be as efficient as possible (because we can't start processing right away from the oldest thing we need).

Might be cool to also offer a "forward" iterator that lets you start from some point.

Reply:
I think that can be in a separate PR

Reply:
agree

Reply:
placeholder (may actually be pretty tough because multiple forks at tip): https://github.com/commonwarexyz/monorepo/issues/2083